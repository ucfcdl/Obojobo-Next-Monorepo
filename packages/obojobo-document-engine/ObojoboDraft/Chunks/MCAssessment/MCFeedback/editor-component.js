import React from 'react'

class MCFeedback extends React.Component {
	delete() {
		const editor = this.props.editor
		const change = editor.value.change()
		change.removeNodeByKey(this.props.node.key)

		editor.onChange(change)
	}
	render() {
		return (
			<div className="component obojobo-draft--chunks--mc-assessment--mc-feedback editor-feedback">
				<button className="editor--page-editor--delete-node-button" onClick={() => this.delete()}>
					X
				</button>
				<span className="label" contentEditable={false}>
					Feedback
				</span>
				{this.props.children}
			</div>
		)
	}
}

export default MCFeedback
