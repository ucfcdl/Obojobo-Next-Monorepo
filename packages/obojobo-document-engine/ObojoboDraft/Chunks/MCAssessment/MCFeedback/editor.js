import React from 'react'
import Common from 'Common'

import Node from './editor-component'
import Schema from './schema'
import Converter from './converter'

const MCFEEDBACK_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCFeedback'

const plugins = {
	renderNode(props) {
		switch (props.node.type) {
			case MCFEEDBACK_NODE:
				return <Node {...props} {...props.attributes} />
		}
	},
	schema: Schema
}

Common.Store.registerEditorModel('ObojoboDraft.Chunks.MCAssessment.MCFeedback', {
	name: 'Multiple Choice Feedback',
	isInsertable: false,
	slateToObo: Converter.slateToObo,
	oboToSlate: Converter.oboToSlate,
	plugins
})

const MCFeedback = {
	components: {
		Node
	},
	helpers: {
		slateToObo: Converter.slateToObo,
		oboToSlate: Converter.oboToSlate
	},
	plugins
}

export default MCFeedback
