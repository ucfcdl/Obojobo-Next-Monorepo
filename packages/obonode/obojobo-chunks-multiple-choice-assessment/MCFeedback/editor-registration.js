import React from 'react'
import { Node, Element, Transforms, Text, Editor } from 'slate'
import Common from 'obojobo-document-engine/src/scripts/common'
import NormalizeUtil from 'obojobo-document-engine/src/scripts/oboeditor/util/normalize-util'

import EditorComponent from './editor-component'
import Converter from './converter'

const MCCHOICE_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCChoice'
const MCANSWER_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCAnswer'
const MCFEEDBACK_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCFeedback'
const TEXT_NODE = 'ObojoboDraft.Chunks.Text'

const MCFeedback = {
	name: MCFEEDBACK_NODE,
	menuLabel: 'Multiple Choice Feedback',
	isInsertable: false,
	supportsChildren: true,
	helpers: Converter,
	plugins: {
		// Editor Plugins - These get attached to the editor object and override it's default functions
		// They may affect multiple nodes simultaneously
		normalizeNode(entry, editor, next) {
			const [node, path] = entry

			// If the element is a MCFeedback, only allow Content children
			if (Element.isElement(node) && node.type === MCFEEDBACK_NODE) {
				for (const [child, childPath] of Node.children(editor, path)) {
					if (Element.isElement(child) && !Common.Registry.contentTypes.includes(child.type)) {
						Transforms.removeNodes(editor, { at: childPath })
						return
					}

					// Wrap loose text children in a Text Node
					if (Text.isText(child)) {
						Transforms.wrapNodes(
							editor,
							{
								type: TEXT_NODE,
								content: { indent: 0 }
							},
							{ at: childPath }
						)
						return
					}
				}

				// MCFeedback parent normalization
				// Note - collect up an adjacent MCAnswer, if it exists
				const [parent] = Editor.parent(editor, path)
				if (!Element.isElement(parent) || parent.type !== MCCHOICE_NODE) {
					NormalizeUtil.wrapOrphanedSiblings(
						editor,
						entry,
						{
							type: MCCHOICE_NODE,
							content: {},
							children: []
						},
						node => node.type === MCANSWER_NODE
					)
					return
				}
			}

			next(entry, editor)
		},
		// Editable Plugins - These are used by the PageEditor component to augment React functions
		// They affect individual nodes independently of one another
		renderNode(props) {
			return <EditorComponent {...props} {...props.attributes} />
		}
	}
}

export default MCFeedback
