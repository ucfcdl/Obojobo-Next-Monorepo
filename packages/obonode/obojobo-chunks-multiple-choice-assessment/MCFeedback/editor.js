import React from 'react'
import Common from 'obojobo-document-engine/src/scripts/common'

import Node from './editor-component'
import Schema from './schema'
import Converter from './converter'

const MCFEEDBACK_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCFeedback'

const plugins = {
	renderNode(props, editor, next) {
		switch (props.node.type) {
			case MCFEEDBACK_NODE:
				return <Node {...props} {...props.attributes} />
			default:
				return next()
		}
	},
	schema: Schema
}

Common.Registry.registerModel('ObojoboDraft.Chunks.MCAssessment.MCFeedback', {
	name: 'Multiple Choice Feedback',
	isInsertable: false,
	slateToObo: Converter.slateToObo,
	oboToSlate: Converter.oboToSlate,
	supportsChildren: true,
	plugins
})

const MCFeedback = {
	components: {
		Node
	},
	helpers: {
		slateToObo: Converter.slateToObo,
		oboToSlate: Converter.oboToSlate
	},
	plugins
}

export default MCFeedback
