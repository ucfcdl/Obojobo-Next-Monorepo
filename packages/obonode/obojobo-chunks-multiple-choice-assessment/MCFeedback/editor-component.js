import './viewer-component.scss'
import './editor-component.scss'

import React from 'react'
import Common from 'obojobo-document-engine/src/scripts/common'

const { Button } = Common.components

class MCFeedback extends React.Component {
	delete() {
		const editor = this.props.editor
		return editor.removeNodeByKey(this.props.node.key)
	}
	render() {
		return (
			<div className="component obojobo-draft--chunks--mc-assessment--mc-feedback editor-feedback">
				<Button className="delete-button" onClick={this.delete.bind(this)}>
					Ã—
				</Button>
				<span className="label" contentEditable={false}>
					Feedback
				</span>
				{this.props.children}
			</div>
		)
	}
}

export default MCFeedback
