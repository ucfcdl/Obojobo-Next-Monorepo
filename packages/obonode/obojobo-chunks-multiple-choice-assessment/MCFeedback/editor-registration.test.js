import { Transforms } from 'slate'
import React from 'react'
import NormalizeUtil from 'obojobo-document-engine/src/scripts/oboeditor/util/normalize-util'
jest.mock('obojobo-document-engine/src/scripts/oboeditor/util/normalize-util')

import MCFeedback from './editor-registration'

const MCCHOICE_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCChoice'
const MCFEEDBACK_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCFeedback'
const TEXT_NODE = 'ObojoboDraft.Chunks.Text'

jest.mock('obojobo-document-engine/src/scripts/common', () => ({
	Registry: {
		contentTypes: ['ObojoboDraft.Chunks.Text']
	},
	components: {
		// eslint-disable-next-line react/display-name
		Button: props => <button {...props}>{props.children}</button>
	}
}))
jest.mock('./converter', () => ({}))

describe('MCFeedback editor', () => {
	test('plugins.renderNode renders a node', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: MCFEEDBACK_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(MCFeedback.plugins.renderNode(props, null, jest.fn())).toMatchSnapshot()
	})

	test('normalizeNode calls next if the node is not a MCFeedback node', () => {
		const next = jest.fn()
		MCFeedback.plugins.normalizeNode([{}, []], {}, next)

		expect(next).toHaveBeenCalled()
	})

	test('normalizeNode on MCFeedback calls next if all MCFeedback children are valid', () => {
		const next = jest.fn()
		const editor = {
			children: [
				{
					id: 'mockKey',
					type: MCCHOICE_NODE,
					content: {},
					children: [
						{
							id: 'mockKey',
							type: MCFEEDBACK_NODE,
							content: {},
							children: [
								{
									type: TEXT_NODE,
									content: { indent: 1 },
									children: [{ text: 'mockCode', b: true }]
								}
							]
						}
					]
				}
			],
			isInline: () => false
		}
		MCFeedback.plugins.normalizeNode([editor.children[0].children[0], [0, 0]], editor, next)

		expect(next).toHaveBeenCalled()
	})

	test('normalizeNode on MCFeedback calls Transforms on invalid Element children', () => {
		jest.spyOn(Transforms, 'removeNodes').mockReturnValueOnce(true)

		const next = jest.fn()
		const editor = {
			children: [
				{
					id: 'mockKey',
					type: MCCHOICE_NODE,
					content: {},
					children: [
						{
							id: 'mockKey',
							type: MCFEEDBACK_NODE,
							content: {},
							children: [
								{
									type: 'improperNode',
									content: { indent: 1 },
									children: [{ text: 'mockCode', b: true }]
								}
							]
						}
					]
				}
			],
			isInline: () => false
		}
		MCFeedback.plugins.normalizeNode([editor.children[0].children[0], [0, 0]], editor, next)

		expect(Transforms.removeNodes).toHaveBeenCalled()
	})

	test('normalizeNode on MCFeedback calls Transforms on invalid Text children', () => {
		jest.spyOn(Transforms, 'wrapNodes').mockReturnValueOnce(true)

		const next = jest.fn()
		const editor = {
			children: [
				{
					id: 'mockKey',
					type: MCCHOICE_NODE,
					content: {},
					children: [
						{
							id: 'mockKey',
							type: MCFEEDBACK_NODE,
							content: {},
							children: [{ text: 'mockCode', b: true }]
						}
					]
				}
			],
			isInline: () => false
		}
		MCFeedback.plugins.normalizeNode([editor.children[0].children[0], [0, 0]], editor, next)

		expect(Transforms.wrapNodes).toHaveBeenCalled()
	})

	test('normalizeNode on MCFeedback calls Transforms with invalid parent', () => {
		const next = jest.fn()
		const editor = {
			children: [
				{
					id: 'mockKey',
					type: MCFEEDBACK_NODE,
					content: {},
					children: [
						{
							type: TEXT_NODE,
							content: { indent: 1 },
							children: [{ text: 'mockCode', b: true }]
						}
					]
				}
			],
			isInline: () => false
		}
		NormalizeUtil.wrapOrphanedSiblings.mockImplementation((editor, entry, wrapper, match) => {
			match(editor.children[0])
		})
		MCFeedback.plugins.normalizeNode([editor.children[0], [0]], editor, next)

		expect(NormalizeUtil.wrapOrphanedSiblings).toHaveBeenCalled()
	})
})
