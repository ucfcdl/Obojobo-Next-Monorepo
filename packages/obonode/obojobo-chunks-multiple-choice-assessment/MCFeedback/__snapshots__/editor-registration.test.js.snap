// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MCFeedback editor plugins.renderNode calls next 1`] = `undefined`;

exports[`MCFeedback editor plugins.renderNode renders a node 1`] = `
<MCFeedback
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Chunks.MCAssessment.MCFeedback",
    }
  }
/>
`;
